cdap-ui/app/cdap/components/AbstractWidget/FunctionDropdownAliasWidget/FuctionDropdownRow.tsx

```diff
@@ -0,0 +1,194 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import Input from '@material-ui/core/Input';
+import withStyles, { StyleRules } from '@material-ui/core/styles/withStyles';
+import Select from '@material-ui/core/Select';
+import MenuItem from '@material-ui/core/MenuItem';
+import AbstractRow, {
+  IAbstractRowProps,
+  AbstractRowStyles,
+} from 'components/AbstractWidget/AbstractMultiRowWidget/AbstractRow';
+
+const styles = (theme): StyleRules => {
+  return {
+    ...AbstractRowStyles(theme),
+    inputContainer: {
+      display: 'grid',
+      gridTemplateColumns: '1fr 1fr 30px 1fr',
+      gridGap: '10px',
+    },
+    disabled: {
+      color: `${theme.palette.grey['50']}`,
+    },
+    separator: {
+      textAlign: 'center',
+    },
+  };
+};
+
+interface IComplexDropdown {
+  value: string | number;
+  label: string;
+}
+
+export type IDropdownOption = string | number | IComplexDropdown;
+
+interface IFunctionDropdownRowProps extends IAbstractRowProps<typeof styles> {
+  placeholders: Record<string, string>;
+  dropdownOptions: IDropdownOption[];
+}
+
+interface IKeyValueState {
+  field: string;
+  func: string;
+  alias: string;
+}
+
+type StateKeys = keyof IKeyValueState;
+
+class FunctionDropdownRow extends AbstractRow<IFunctionDropdownRowProps, IKeyValueState> {
+  public static defaultProps = {
+    placeholders: {
+      field: 'field',
+      alias: 'alias',
+    },
+    dropdownOptions: [],
+  };
+
+  public state = {
+    field: '',
+    func: '',
+    alias: '',
+  };
+
+  /**
+   * Sample input: fieldName:Avg(alias)
+   */
+  public componentDidMount() {
+    const [alias, fn] = this.props.value.split(':');
+    const { func, field } = this.extractFunctionAndAlias(fn);
+
+    this.setState({
+      field,
+      func,
+      alias,
+    });
+  }
+
+  private extractFunctionAndAlias(fn) {
+    const defaultResponse = {
+      func: '',
+      field: '',
+    };
+
+    if (!fn) {
+      return defaultResponse;
+    }
+
+    const openBracketIndex = fn.indexOf('(');
+    const closeBracketIndex = fn.indexOf(')');
+
+    if (openBracketIndex === -1 || closeBracketIndex === -1) {
+      return defaultResponse;
+    }
+
+    return {
+      func: fn.substring(0, openBracketIndex),
+      field: fn.substring(openBracketIndex + 1, closeBracketIndex),
+    };
+  }
+
+  private handleChange = (type: StateKeys, e) => {
+    this.setState(
+      {
+        [type]: e.target.value,
+      } as Pick<IKeyValueState, StateKeys>,
```

Comment:
This is strange. Even though we define the state and `type` we still need to do this type assertion to pick keys from `IKeyValueState` and `StateKeys`?

Reply:
yeah, somehow it cannot detect it... :(